### ISSUE TITLE
Race Condition in TrackAsyncAction 

### ISSUE THREAD

**Issue by smj389** [Microsoft/Win2D#585]
In TrackAsyncAction in RecreatableDeviceManager.impl.h , there is a race condition. It is possible that the user supplied IAsyncAction may have entered status completed before the completed handler is registered (line 549). The code unconditionally sets m_currentOperationisPending to true without checking whether the action has already completed. I noticed this when setting up some code with an empty create_resource_async function. In this case, the draw and update events are never raised by the CanvasAnimatedControl.

**Comment by damyanp** [comment]
Thanks for reporting this.

I suspect that the fix for this will be to set m_currentOperationIsPending before the call to put_Completed.

**Comment by smj389** [comment]
That was what I thought.

**Comment by smj389** [comment]
On second thought, tt may be worth checking whether the IAsyncAction is completed before registering the handler. If the Status is Started, then register the handler. After put_completed, then check again before setting m_currentOperationIsPending. This way there is no possibility of a missed pulse. In that case, it is possible that the handler will run unnecessarily. But it can notice that m_currentOperationIsPending is false and leave it alone.

**Comment by smj389** [comment]
I don't have time to mock something up now - but I can later if that will help.

**Comment by damyanp** [comment]
I think that checking the Status on the IAsyncInfo would still have the race in it, since there's nothing to prevent the action completing between the call to get_Status and the call to put_Completed.  

The design of IAsyncAction prevents this race, since IAsyncAction implementations are expected to call the completion handler if put_Completed is called on them if they're in a terminal state.

**Comment by smj389** [comment]
You are right. I hadn't realised that the mutex guarding m_currentOperationIsPending was recursive. If it hadn't been, I guess it would have dead-locked when the completed handler was instantly invoked by put_Completed on the same thread.
